third_party/move/move-vm/types/src/delayed_values/delayed_field_id.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    delayed_values::{
+        bcs_utils,
+        derived_string_snapshot::{
+            bytes_and_width_to_derived_string_struct, derived_string_struct_to_bytes_and_length,
+            from_utf8_bytes, is_derived_string_struct_layout, u128_to_u64,
+            u64_to_fixed_size_utf8_bytes,
+        },
+    },
+    values::{Struct, Value},
+};
+use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_core_types::{value::MoveTypeLayout, vm_status::StatusCode};
+
+/// Represents a unique 32-bit identifier used for values which also stores their
+/// serialized size (u32::MAX at most). Can be stored as a single 64-bit unsigned
+/// integer.
+const BITS_FOR_SIZE: usize = 32;
+
+/// Ephemeral identifier type used by delayed fields (aggregators, snapshots)
+/// during execution.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct DelayedFieldID {
+    unique_index: u32,
+    // Exact number of bytes serialized delayed field will take.
+    width: u32,
+}
+
+impl DelayedFieldID {
+    pub fn new_with_width(unique_index: u32, width: u32) -> Self {
+        Self {
+            unique_index,
+            width,
+        }
+    }
+
+    pub fn new_for_test_for_u64(unique_index: u32) -> Self {
+        Self::new_with_width(unique_index, 8)
+    }
+
+    pub fn as_u64(&self) -> u64 {
+        ((self.unique_index as u64) << BITS_FOR_SIZE) | self.width as u64
+    }
+
+    pub fn into_derived_string_struct(self) -> PartialVMResult<Value> {
+        let width = self.extract_width() as usize;
+
+        // we need to create DerivedString struct that serializes to exactly match given `width`.
+        // I.e: size_u32_as_uleb128(value.len()) + value.len() + size_u32_as_uleb128(padding.len()) + padding.len() == width
+        // As padding has a fixed allowed max width, it is easiest to expand value to have the padding be minimal.
+        // We cannot always make padding to be 0 byte vector (serialized into 1 byte) - as not all sizes are possible
+        // for string due to variable encoding of string length.
+
+        // So we will over-estimate the serialized length of the value a bit.
+        let value_len_width_upper_bound = bcs_utils::size_u32_as_uleb128(width - 2); // we subtract 2 because uleb sizes (for both value and padding fields) are at least 1 byte.
+
+        // If we don't even have enough space to store the length of the value, we cannot proceed
+        if width <= value_len_width_upper_bound + 1 {
+            return Err(PartialVMError::new(StatusCode::DELAYED_FIELDS_CODE_INVARIANT_ERROR).with_message(format!(
```

Comment:
why did you remove code_invariant_error function? I don't want any direct creation of DELAYED_FIELDS_CODE_INVARIANT_ERROR, because then nothing gets logged. we log on the call to itself, and that is what we will alert on

(it was private and duplicate, because logger is not available here, so we just print, for those calls inside this file)

Reply:
`code_invariant_error`/logger is in aptos types?

Reply:
done